src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java

```diff
@@ -691,15 +700,22 @@ public void test_clone_refspecs() throws Exception {
       w.git.clone_().url(localMirror()).refspecs(refspecs).repositoryName("origin").execute();
       w.git.withRepository((Repository repo, VirtualChannel channel) -> {
           String[] fetchRefSpecs = repo.getConfig().getStringList(ConfigConstants.CONFIG_REMOTE_SECTION, Constants.DEFAULT_REMOTE_NAME, "fetch");
-          assertEquals("Expected 2 refspecs", 2, fetchRefSpecs.length);
-          assertEquals("Incorrect refspec 1", "+refs/heads/master:refs/remotes/origin/master", fetchRefSpecs[0]);
-          assertEquals("Incorrect refspec 2", "+refs/heads/1.4.x:refs/remotes/origin/1.4.x", fetchRefSpecs[1]);
+          int refSpecIndex = 0;
+          if( fetchRefSpecs.length == 3 ) {
```

Comment:
This will cause a major increase in repository size for those repositories which previously relied on the refspec being honored on first fetch.

By adding the "full width" refspec, the other two refspecs which follow are extraneous.  They are already covered by the full width refspec.

Reply:
As specified in the pull request description above, git clone implicitly adds the full width refspec when doing the initial git clone. According to the git documentation there is no possibility to limit the refspec during git clone: https://git-scm.com/docs/git-clone

According to our experience cloning a repository with the full refspec but without checkout (--no-checkout) and then checking out the requested ref is much faster than fetching only the required refspec initiallly. 

I understand your concerns regarding the repository size for projects that used a specific refspec to save space while sacrificing it for the slower performance of git fetch. This prohibits changing the default behaivour of the clone command. Would it be acceptable to have this behaviour configurable and off by default?

Reply:
From the discussion in https://issues.jenkins-ci.org/browse/JENKINS-23345 there seem to be at least some other people that have similar performance issues with git init+fetch in comparison to git clone.

Reply:
You asked:

> Would it be acceptable to have this behaviour configurable and off by default?

I think you're asking if it would be acceptable to allow the caller of the git client API to optionally choose if it should use clone rather than init and fetch.  Yes, that would be acceptable, so long as the default remained to use init and fetch.  There are more than 130000 Jenkins installations that depend on the git plugin.  Many of those installations use it in ways that are critical to their production environments.

We very rarely change default behaviors of the git client plugin or the git plugin because so many installations depend on those default behaviors.

Reply:
You mentioned JENKINS-23345 but didn't mention the performance comparisons I reported as part of that bug report.  Can you share the commands you're using and the `time` output of those commands so that I have some idea why `clone` is significantly faster than `init+fetch` in your case?

I understand from the comments by Jason Salaz of GitHub that there are cases where clone will be significantly faster than init + fetch.  Unfortunately, I was unable to duplicate those cases.  I was able to duplicate that `init+fetch` is slightly slower.  I don't see any way to retain the existing git client plugin behaviors and still switch to `clone`.  

If the implementation switches to `git clone` then narrow refspecs are out (since git clone doesn't take a refspec as an argument).

It should be possible in the implementation you provided to use a reference repository from the clone command.

I think many of the narrow refspec cases might be supported by newer command line git implementations if the `--single-branch --branch BRANCH_NAME` arguments are passed.  Not all of them, but many of the cases that I've encountered could be handled by that.

If a parsing of the user provided refspec (if one is set) shows that it could only match a single branch name, then `git clone --single-branch --branch BRANCH_NAME` has the same performance as `init+fetch` .  In my example, they both complete in less than 1 second when the clone with full width refspec requires 8 or more seconds